### PR TITLE
test: add dns resolver tests

### PR DESCRIPTION
**Pull request by tynes**
This PR includes a test that purchases names and then writes one of each of the different Handshake records into the tree. A DNS stub resolver then queries for the name and asserts on the correct data being returned.

The DNSSEC test tests that all resource records sets in the answer and authority sections are signed. If all resource record sets in the additional section were signed as well, it could cause problems because there could potentially be many signatures and would force the DNS server to fall back to TCP. Ideally a different authentication mechanism is devised, because the DNSSEC key is public information anyways.

@boymanjor happy to split this up into smaller PRs if it would make it easier to review

Closes https://github.com/handshake-org/hsd/issues/264

### CODE REVIEW

File: lib/dns/records.js [hunk 1]
```diff
@@ -64,14 +64,72 @@ const types = {
...
+  [types.ADDR]: 'ADDR'
+};
+
+/**
+ * DNS RR Types by Handshake
+ * record type.
+ * @constant
+ */
+
+const rrTypeByType = {
```
Comment (tynes) [inline lib/dns/records.js]: Nit: `rrTypesByType`

Also open to an alternative name to this if there are any suggestions.

File: lib/dns/resource.js [hunk 2]
```diff
@@ -1125,6 +1125,8 @@ class Resource extends Struct {
             for (const rr of this.toGlue(types.SRV))
               res.additional.push(rr);
             key.signZSK(res.answer, types.SRV);
+            key.signZSK(res.additional, types.A);
+            key.signZSK(res.additional, types.AAAA);
```
Comment (tynes) [inline lib/dns/resource.js]: `key.signZSK` will throw an error if there are 2 RRs of the same type with a different name. Depending on `toSRVIP` and `toGlue`, I think this is possible. Removing DNSSEC from the `additional` section is one option, or handling that case in `key.signZSK` is another option. Down the codepath of `key.signZSK`, `bns.util.extractSet` is called without a name.

File: lib/dns/resource.js [hunk 3]
```diff
@@ -1224,13 +1228,27 @@ class Resource extends Struct {
...
         key.signZSK(res.additional, types.AAAA);
         break;
-      case types.NS:
+      case types.NS: {
         res.authority = this.toNS(name);
         res.additional = this.toNSIP(name);
         for (const rr of this.toGlue(types.NS))
           res.additional.push(rr);
         key.signZSK(res.authority, types.NS);
+
```
Comment (tynes) [inline lib/dns/resource.js]: This logic handles the case when there are multiple names of the same RR type. I think this logic should move inside of `key.signZSK` so that the abstraction here can stay the same. A single `key.signZSK` would be able to append multiple `RRSIG` records if there are multiple rrsets in the same section, ie they have different names but the same rr type (since that is the second argument to `key.signZSK`.

File: lib/node/rpc.js [hunk 4]
```diff
@@ -2366,7 +2366,7 @@ class RPC extends RPCBase {
 
     try {
       const res = Resource.decode(ns.data);
-      return res.toJSON();
```
Comment (tynes) [inline lib/node/rpc.js]: This change is required to correctly render the JSON response with the `name` field filled.

File: lib/dns/records.js (hunk 1 above)
Comment (tynes) [reply lib/dns/records.js]: `hnsTodns`

File: lib/dns/records.js [hunk 5]
```diff
@@ -850,8 +908,8 @@ class SMIME extends Struct {
     assert((json.usage & 0xff) === json.usage);
     assert((json.selector & 0xff) === json.selector);
     assert((json.matchingType & 0xff) === json.matchingType);
-    assert(typeof json.fingerprint === 'string');
-    assert((json.fingerprint.length >>> 1) <= 255);
+    assert(typeof json.certificate === 'string');
```
Comment (tynes) [inline lib/dns/records.js]: TODO: break this into its own PR

File: lib/dns/server.js [hunk 6]
```diff
@@ -48,17 +50,6 @@ const {
 const TYPE_MAP = Buffer.from('000722000000000380', 'hex');
 const RES_OPT = { inet6: false, tcp: true };
 
-const blacklist = new Set([
```
Comment (tynes) [inline lib/dns/server.js]: TODO: break this into its own PR

File: lib/node/rpc.js (hunk 4 above)
Comment (tynes) [reply lib/node/rpc.js]: TODO: break this into its own PR

File: lib/dns/records.js (hunk 5 above)
Comment (tynes) [reply lib/dns/records.js]: Merged in https://github.com/handshake-org/hsd/commit/0da753980c0899667d52188883ccd728f9355e96

File: lib/dns/server.js (hunk 6 above)
Comment (tynes) [reply lib/dns/server.js]: Closed, considering different options. Remove this commit from this PR

File: lib/node/rpc.js (hunk 4 above)
Comment (tynes) [reply lib/node/rpc.js]: Merged in https://github.com/handshake-org/hsd/commit/a90a45bb9ea2c267e47bbaa173da1f181123f35f

File: lib/dns/records.js [hunk 7]
```diff
@@ -64,14 +64,72 @@ const types = {
...
+  [types.PGP]: 'PGP',
+  [types.ADDR]: 'ADDR'
+};
+
+/**
+ * DNS RR Types by Handshake
+ * record type.
+ * @constant
+ */
+
```
Comment (tynes) [inline lib/dns/records.js]: Not sure what to call this data structure. It is used to dynamically map between Handshake record type and DNS record type. Any feedback would be appreciated.

Comment (tynes) [reply lib/dns/records.js]: @boymanjor suggested `hnsToDns`

File: test/dns-test.js [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+    await nclient.execute('generatetoaddress', [1, address]);
+    await common.forValue(obj, 'complete', true);
+  }
+}
+
+/**
+ * Verify DNSSEC
+ */
+
+function verifyDNSSEC(resource, pubkey, qtype, name) {
```
Comment (tynes) [inline test/dns-test.js]: Refactor this to only work on answer, potentially authority.

File: lib/dns/resource.js (hunk 2 above)
Comment (tynes) [reply lib/dns/resource.js]: Don't sign the additional section here

File: test/dns-test.js (hunk 8 above)
Comment (tynes) [reply test/dns-test.js]: This has been completed

File: test/dns-test.js [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+      additional.data.signature = null;
+  }
+}
+
+/**
+ * Parse the ZSK and KSK from
+ * a record.
+ */
+
+function getSigningKeys(record) {
```
Comment (tynes) [inline test/dns-test.js]: It is possible to do this without calling `rr.toJSON` and looking at the `keyType`, but it is a lot more simple to do it this way when performance isn't critical.

File: test/dns-test.js [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+
+  return {
+    zsk: zsk,
+    ksk: ksk
+  };
+}
+
+/**
+ * Build a name to query for the types
+ * that require additional labels.
```
Comment (tynes) [inline test/dns-test.js]: I'd like to turn this into a helper function as I think I'd find it useful building applications on top of Handshake. Not quite sure where it would live or how to generalize it correctly - notice that it uses a magic index into an array.

File: lib/dns/resource.js (hunk 2 above)
Comment (tynes) [reply lib/dns/resource.js]: > Don't sign the additional section here

This has been applied, if the additional section is signed then the response will be too large and fall back to TCP. The key is public anyways.

For more, see:
https://security.stackexchange.com/questions/91038/dnssec-why-are-rrsigs-returned-in-the-answer-not-additional-section

File: lib/dns/resource.js (hunk 3 above)
Comment (tynes) [reply lib/dns/resource.js]: This is probably good to note for the future, but is no longer relevant.

File: test/dns-test.js [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+    await node.close();
+    await nstub.close();
+    await rstub.close();
+  });
+
+  describe('Recursive Resolver', () => {
+    it('should return an A record', async () => {
+      const name = 'google.com.';
+
+      const res = await rstub.lookup(name, types.A);
```
Comment (pinheadmz) [inline test/dns-test.js]: Is this the first time a test in bcoin or hsd actually makes an external internet connection? I wonder what the implications of that are, and what kind of failures could happen with CI. I think `google.com` seems like it will be a safe bet to query for a long time, but it still makes this test rely on an external factor. Do we want to consider a separate testing script like what I did here: https://github.com/bcoin-org/bcoin/pull/759 -- putting the test that requires an internet connection outside the main testing suite, but still running it on CI. Just 🍕 for 🤔 .

Comment (tynes) [reply test/dns-test.js]: Would agree that removing it is probably a good idea, I added it as the first test in this file to figure out how the APIs work together. Thanks for the feedback. It is a risk of nondeterminism, but if `google.com` doesn't resolve (assuming no CI firewall) there is a major problem happening with the internet

File: lib/dns/records.js (hunk 7 above)
Comment (tynes) [reply lib/dns/records.js]: I haven't been able to find a better name for this data structure and have began to like its current name. It is very descriptive of what it contains.

File: test/dns-test.js (hunk 11 above)
Comment (pinheadmz) [reply test/dns-test.js]: Both `"should return no answer..."` tests also hit the internet. Can be seen by logging the queries in `bns/lib/resolver/dns.js` `exchange()`

Comment (pinheadmz) [reply test/dns-test.js]:
```
Querying server: 127.0.0.1 (43708) (tcp=false)
[ { name: 'com.', class: 'IN', type: 'A' } ]
Querying server: 192.112.36.4 (18481) (tcp=false)
[ { name: 'com.', class: 'IN', type: 'NS' } ]
      ✓ should return no answer for an ICANN rooted name (125ms)
Querying server: 127.0.0.1 (47554) (tcp=false)
[ { name: 'foo.', class: 'IN', type: 'A' } ]
Querying server: 202.12.27.33 (24311) (tcp=false)
[ { name: 'foo.', class: 'IN', type: 'NS' } ]
      ✓ should return no answer for HNS name pre-update (195ms)
```

File: test/dns-test.js [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+    it('should return no answer for an ICANN rooted name', async () => {
+      const name = 'com.';
+
+      const res = await nstub.lookup(name, types.A);
+
+      assert.equal(res.question[0].name, name);
+      assert.equal(res.answer.length, 0);
+
+      // Handshake zone is not authoritative for ICANN based name
+      assert.equal(res.aa, false);
```
Comment (pinheadmz) [inline test/dns-test.js]: Should we test that the nameserver returned an `authority` section from the ICANN lookup?

File: lib/dns/records.js (hunk 7 above)
Comment (tynes) [reply lib/dns/records.js]: `dnsTypeToType`

File: test/dns-test.js [hunk 13]
```diff
@@ -0,0 +1,396 @@
...
+
+    // record to save in the authenticated tree
+    for (const [hstype, item] of Object.entries(json)) {
+      const type = types[item.type];
+      const resource = Resource.fromJSON(item.resource);
+
+      // args: size, height, network
+      const name = rules.grindName(5, 1, network);
+
+      it(`should update ${name}`, async () => {
```
Comment (pinheadmz) [inline test/dns-test.js]: I don't think the `${name}` is useful here since it's randomly generated, and each pass of the test will be different.
Perhaps "should update name with record type ${item.type}"

File: test/dns-test.js [hunk 14]
```diff
@@ -0,0 +1,396 @@
...
+
+        const info = await nclient.execute('getnameinfo', [name]);
+        const data = Buffer.from(info.info.data, 'hex');
+
+        const returned = Resource.decode(data);
+        assert.deepEqual(returned.getJSON(name), json);
+
+        // this commits the tree state to disk
+        // updates served by the dns server
+        await mineBlocks(treeInterval, coinbase);
```
Comment (pinheadmz) [inline test/dns-test.js]: There's a few places around the testing suite where a complete auction process must be run, I think there's also one that uses RPC calls like this. Anyway, that code is _covered_, and is a bit of a waste here, especially since every one of these tests takes ~500ms on my machine. 

I think we should find a simple-unit-test way to just inject a single record into the tree without any of the blockchain BS. This entire test could be 90% lighter. Maybe it's possible to even drop the `FullNode` entirely and just run the StubResolvers, RootServer, and NameServer, with a hacked-in urkel tree.

Comment (pinheadmz) [reply test/dns-test.js]: Looks like `FullNode` instantiates a `RootResolver` with the option:
`lookup: key => this.chain.db.tree.get(key)`
I wonder if that can just be stubbed with a simple map. All the updates in this test could be simple insertions!

File: test/dns-test.js [hunk 15]
```diff
@@ -0,0 +1,396 @@
...
+        const response = await nstub.lookup(query, type);
+        const res = response.toJSON();
+        const dns = resource.toDNS(util.fqdn(name), type).toJSON();
+
+        // query for the zone signing key
+        const dnskey = await nstub.lookup('.', types.DNSKEY);
+
+        // parse the zone signing key and
+        // key signing key out of the response
+        const {zsk, ksk} = getSigningKeys(dnskey);
```
Comment (pinheadmz) [inline test/dns-test.js]: Do these need to run on each loop? Or can they be hoisted

File: test/dns-test.js [hunk 16]
```diff
@@ -0,0 +1,396 @@
...
+
+        assert(zsk instanceof wire.Record);
+        assert(ksk instanceof wire.Record);
+
+        // validate the signature on the ZSK
+        verifyDNSSEC(dnskey, ksk, types.DNSKEY, '.');
+        // validate the signature over the rrsets
+        verifyDNSSEC(response, zsk, type, query);
```
Comment (pinheadmz) [inline test/dns-test.js]: OK I see you are testing the keys here, maybe it should be a separate `it` block, or even run in `before()`

File: test/dns-test.js [hunk 17]
```diff
@@ -0,0 +1,396 @@
...
+/**
+ * Build a name to query for the types
+ * that require additional labels.
+ */
+
+function buildQuery(name, resource, type) {
+  switch (type) {
+    case types.SRV: {
+      const service = resource.service[0];
+      return '_' + service.service
```
Comment (tynes) [inline test/dns-test.js]: Nit: fix the style here to match the other cases, 1 item per line

File: test/dns-test.js (hunk 16 above)
Comment (tynes) [reply test/dns-test.js]: It is testing that the zsk was signed by the ksk. I wouldn't trust that the zsk is correct until I know that it has been signed by the ksk, so that is why I test that it is signed each time. In actuality, it is not possible for the zsk to be man in the middle'ed in this test, so it could be done once. In practice, each request should test this, and as end to end tests I'd like to emulate as close as possible what would happen in reality. If it were more computationally complex I'd definitely agree.

If you ask me to refactor it so that it doesn't bid on a new name each time but instead uses a single name and just updates that name, I'd be happy to do that refactor. It slows down the tests quite a bit, but I didn't want any side effects from previous tests to have a role.

File: test/data/resources-v0.json [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+          "certificate": "0102348790"
+        }],
+        "ttl": 64
+      }
+    },
+    "SMIME": {
+      "type": "SMIMEA",
+      "resource": {
+        "smime": [{
+"hash": "92faa0601d44a3c2b98318d8f7e4cd3b17f25caa6753cc04365d753d",
```
Comment (pinheadmz) [inline test/data/resources-v0.json]: nit: indentation?

File: test/data/resources-v0.json [hunk 19]
```diff
@@ -0,0 +1,200 @@
...
+          "hash": "0e9baa5e5ce8cffd7bf13c37423957ed5e0061c9c8807628979ef475",
+          "publicKey": "b796a94d827d3b3142dde13d945a0467b5c9afa53ae551bdf898b1f3101ef883"
+        }],
+        "ttl": 64
+      }
+    },
+    "ADDR": {
+      "type": "URI",
+      "resource": {
+        "addr": ["bitcoin:bc1qdsr2022f90pum2ptd8ftp4nvpgdhk656ue33qz"],
```
Comment (pinheadmz) [inline test/data/resources-v0.json]: Recommend using a regtest address for this, who knows... This address has history in the blockchain, belongs to somebody.

File: test/data/resources-v0.json (hunk 18 above)
Comment (tynes) [reply test/data/resources-v0.json]: Good catch, must have missed it in my editor when it overflowed to the next line

File: test/dns-test.js (hunk 11 above)
Comment (tynes) [reply test/dns-test.js]: Going to remove all external requests. Thanks for the feedback

File: test/dns-test.js (hunk 14 above)
Comment (tynes) [reply test/dns-test.js]: Going to replace the whole auction process with just writing to the tree directly. This will speed up the tests, remove the need for a bunch of code and better decouple the tests. Good feedback

File: test/dns-test.js (hunk 9 above)
Comment (tuxcanfly) [reply test/dns-test.js]: Might be a silly question but in cases where there's multiple DNSKEY records (and multiple ZSK/KSK for some reason), this would select the last. Is that fine or maybe we could add a `break`?

File: test/dns-test.js [hunk 20]
```diff
@@ -0,0 +1,396 @@
...
+        // when the native backend is being used
+        // because it uses OpenSSL which does not
+        // yet use RFC 6979, so nullify the
+        // signatures before comparing them
+
+        if (bcrypto.native === 2)
+          assert.deepEqual(nullSig(res), nullSig(dns));
+        else
+          assert.deepEqual(res.name, dns.name);
+      });
```
Comment (tuxcanfly) [inline test/dns-test.js]: shouldn't this be `assert.deepEqual(res, dns);`?

File: test/data/resources-v0.json [hunk 21]
```diff
@@ -0,0 +1,200 @@
...
+        "ttl": 64
+      }
```
Comment (tuxcanfly) [inline test/data/resources-v0.json]: All these test cases have only one host item. Would multiple be a good test case?

File: test/dns-test.js [hunk 22]
```diff
@@ -0,0 +1,396 @@
...
+        verifyDNSSEC(dnskey, ksk, types.DNSKEY, '.');
+        // validate the signature over the rrsets
+        verifyDNSSEC(response, zsk, type, query);
+
+        // NOTE: the signatures are not canonical
+        // when the native backend is being used
+        // because it uses OpenSSL which does not
+        // yet use RFC 6979, so nullify the
+        // signatures before comparing them
+
```
Comment (tuxcanfly) [inline test/dns-test.js]: nit: new line after comments

File: test/dns-test.js [hunk 23]
```diff
@@ -0,0 +1,396 @@
...
+        assert(zsk instanceof wire.Record);
+        assert(ksk instanceof wire.Record);
+
+        // validate the signature on the ZSK
+        verifyDNSSEC(dnskey, ksk, types.DNSKEY, '.');
+        // validate the signature over the rrsets
+        verifyDNSSEC(response, zsk, type, query);
+
+        // NOTE: the signatures are not canonical
```
Comment (tuxcanfly) [inline test/dns-test.js]: does this mean `verifyDNSSEC` manages to verify non-canonical signatures? If yes, is it possible to convert to canonical or otherwise verify them instead of skipping them altogether (like nullSig is doing now I presume).

File: lib/dns/records.js (hunk 7 above)
Comment (tuxcanfly) [reply lib/dns/records.js]: indicating that values are DNS types is a good idea. `hnsToDnsTypes` or similar might be good.

File: test/dns-test.js (hunk 20 above)
Comment (tynes) [reply test/dns-test.js]: I can do that but would need to set the edns field along with a few bits in the header. You can see here when the tests run, there are additional fields that are set by the DNS server itself that are not set when calling `resource.toDNS`. Trimmed for brevity. This would expand the test coverage to include those fields being used by the DNS server.

```
  DNS Servers
    Authoritative Resolver
      ✓ should return no answer for HNS name pre-update
      1) should return authenticated INET4 record

  2 passing (948ms)
  20 failing

  1) DNS Servers
       Authoritative Resolver
         should return authenticated INET4 record:

      AssertionError [ERR_ASSERTION]: [Object] deepStrictEqual [Object]

      Expected values to be strictly deep-equal:
      
      + actual - expected ... Lines skipped
      
        {
          aa: true,
      ...
          authority: [],
      +   cd: true,
      -   cd: false,
          code: 'NOERROR',
          code: 'NOERROR',
      +   edns: {
      +     code: 0,
      +     dnssec: true,
      +     enabled: true,
      +     options: [],
      +     size: 4096,
      +     version: 0
      ...
      -   edns: undefined,
      -   id: 0,
      -   opcode: 'QUERY',
      -   qr: false,
      -   question: [],
      -   ra: false,
      -   rd: false,
      ...
      
      assert.deepEqual(res, dns);
             ^

      at Context.it (/home/tynes/Projects/github.com/handshake-org/hsd/test/dns-test.js:170:18)

```

File: test/data/resources-v0.json (hunk 21 above)
Comment (tynes) [reply test/data/resources-v0.json]: Good thinking. You are correct in the sense that there are many different permutations of records inside of a resource. There are some code paths inside of `resource.toDNS` that these mocks do not cover, like reverse resolution. (Address -> Name possible?)

File: test/dns-test.js (hunk 11 above)
Comment (tynes) [reply test/dns-test.js]: Removed this external requests.

File: test/dns-test.js (hunk 13 above)
Comment (tynes) [reply test/dns-test.js]: These are all written directly to the tree now in a `before` block, so this is on longer relevant.

File: test/dns-test.js (hunk 23 above)
Comment (tynes) [reply test/dns-test.js]: `verifyDNSSEC` does work on non-canonical signatures, the problem was that a `deepEqual` did not work because the signatures were not canonical when using the native backend that wraps openssl.

**Comment by tynes** [comment]
This PR depends on https://github.com/handshake-org/hsd/pull/286/files

File: test/dns-test.js (hunk 12 above)
Comment (tynes) [reply test/dns-test.js]: This code block was refactored out of the test suite

File: test/dns-test.js [hunk 24]
```diff
@@ -0,0 +1,396 @@
...
+      assert.equal(question.name, name);
+
+      const [answer] = res.answer;
+      assert(typeof answer === 'object');
+      assert.equal(answer.name, name);
+    });
+  });
+
+  describe('Authoritative Resolver', () => {
+    it('should return no answer for an ICANN rooted name', async () => {
```
[the remark this answers is not in the document]
Comment (tynes) [reply test/dns-test.js]: > I wonder if this test should be duplicated to simulate the scenario in which `com.` actually claims their name on hns.

I like this idea but think that it would expand the scope of this PR

File: test/data/resources-v0.json (hunk 19 above)
Comment (tynes) [reply test/data/resources-v0.json]: Using both a mainnet and a testnet address that I generated locally

File: test/data/resources-v0.json (hunk 21 above)
Comment (tynes) [reply test/data/resources-v0.json]: Added a new test case that covers multiple. The way that the test file is set up is that the key is the name of the handshake record and there is a single object following, so it is not possible to have multiple test cases per handshake record type. I could refactor the file to be a list for each handshake record type, and the test could iterate through each object in that list. Do you think that is a good idea? @tuxcanfly

**Comment by tynes** [comment]
> Might be a silly question but in cases where there's multiple DNSKEY records (and multiple ZSK/KSK for some reason), this would select the last. Is that fine or maybe we could add a break?

I added an assertion for 1 zsk and 1 ksk being returned. I think that will be the expected behavior into the foreseeable future, as the dnssec keys aren't super important to the security of the network since the private key is public information.

It would be cool if we could allow a config option to set a dnssec key, then each resolver would be able to have their own. I think this is possible because the keys themselves are not kept on chain and are not part of consensus.

File: lib/dns/records.js (hunk 7 above)
Comment (tynes) [reply lib/dns/records.js]: Ended up calling it `dnsByType` and have assertions in the test suite that the values are correct. Every time that we talk about what to call it, we come up with a different name :man_shrugging:

File: test/dns-test.js [hunk 25]
```diff
@@ -0,0 +1,298 @@
...
+        // because it uses OpenSSL which does not
+        // yet use RFC 6979, so nullify the
+        // signatures before comparing them
+        if (bcrypto.native === 2) {
+          dns = nullSig(dns);
+          res = nullSig(res);
+          assert(dns && res);
+        }
+
+        assert.deepEqual(dns.answer, res.answer);
```
Comment (tynes) [inline test/dns-test.js]: @tuxcanfly I am no longer checking equality of the dns message header, only the `answer`, `authority` and `additional`. The rest of the message may be different based on the configuration of the dns resolver/server.

File: test/dns-test.js (hunk 10 above)
Comment (tynes) [reply test/dns-test.js]: I'd like to check through `hsd` and `bns` to make sure that it already doesn't exist. Would the fn signature be:

`name (str), hsd resource (Resource), dns type (int) -> List[str] (list of query strings)`

A Resource can have multiple query strings depending on the Records inside of it, but this also may not return the name number of strings in the return value as the number of Records in the Resource which may be confusing to a library consumer.

File: test/dns-test.js [hunk 26]
```diff
@@ -0,0 +1,298 @@
...
+        ns.set(raw, 0);
+        ns.setData(resource.encode());
+
+        // Create a database transaction, as
+        // writing directly to the database is
+        // discouraged. Write to the txn and
+        // then commit it.
+        const txn = node.chain.db.tree.transaction();
+        await txn.insert(nameHash, ns.encode());
+        await txn.commit();
```
Comment (pinheadmz) [inline test/dns-test.js]: Awesome!

File: test/dns-test.js (hunk 10 above)
Comment (tuxcanfly) [reply test/dns-test.js]: OK if we not maybe `test/util/common` is a good place.

File: test/data/resources-v0.json (hunk 21 above)
Comment (tuxcanfly) [reply test/data/resources-v0.json]: Yeah that would work :+1:

File: test/dns-test.js (hunk 25 above)
Comment (tuxcanfly) [reply test/dns-test.js]: Sounds good! :+1:

File: test/dns-test.js (hunk 23 above)
Comment (tuxcanfly) [reply test/dns-test.js]: OK.

**Comment by tynes** [comment]
Need to rebase and squash these commits

File: test/dns-test.js (hunk 15 above)
Comment (tynes) [reply test/dns-test.js]: We know that technically the `DNSKEY` record shouldn't change, sending a request for it each time is extra paranoia. In practice the `DNSKEY` could be rolled over for a zone. It takes awhile for the root zone key to roll over, I wonder how often other zone keys roll over.

File: test/data/resources-v0.json (hunk 21 above)
Comment (tynes) [reply test/data/resources-v0.json]: This has been resolved

File: test/dns-test.js (hunk 10 above)
Comment (boymanjor) [reply test/dns-test.js]: I think we should leave this as a specific test helper for now. We can abstract this into a proper library method when the functionality is needed elsewhere.

Comment (tynes) [reply test/dns-test.js]: Ok this makes sense to me. Its not clearly documented anywhere in the Handshake docs that you need to use a particular domain name (with subdomains) to query for some of the rr types. Maybe the `bns` stub resolver could be extended for this repo where its wrapped with an easier to use API that will handle these cases as well as use handshake native types instead of dns rr types. I could imagine that being very helpful for onboarding new developers, especially if it works in the browser.

File: lib/dns/resource.js [hunk 27]
```diff
@@ -1203,10 +1203,12 @@ class Resource extends Struct {
       case types.A:
         res.answer = this.toA(name);
         key.signZSK(res.answer, types.A);
+        key.signZSK(res.answer, types.TXT);
```
[the remark this answers is not in the document]
Comment (tynes) [reply lib/dns/resource.js]: Opened a PR here: https://github.com/handshake-org/hsd/pull/292

Note that the test coverage for this is actually in this PR. I will rebase this PR on top of `master` once that is merged

**Comment by tynes** [comment]
> I also think we can clean up the commit history.

I can rebase on master and rebase out a bunch of the intermediate commits after https://github.com/handshake-org/hsd/pull/292 gets merged.

Thanks for the review @boymanjor 

Note that I wanted to add SIG0 validation to this PR as well, I started in https://github.com/handshake-org/hsd/pull/289 but ran into a bug with `KEY` records. Right now I do not see a way to validate the SIG0 signature without being able to query for the key, which the [RFC](https://tools.ietf.org/html/rfc2931) defines `KEY` as the RR type used to return the SIG0 key. The authoritative resolver doesn't have logic for `KEY` RR types at the name `.`

See: https://github.com/handshake-org/hsd/blob/master/lib/dns/server.js#L228

File: lib/dns/resource.js (hunk 27 above)
[the remark this answers is not in the document]
Comment (boymanjor) [reply lib/dns/resource.js]: @tynes #292 has been merged.

**Comment by tynes** [comment]
Needs rebase now that https://github.com/handshake-org/hsd/commit/91f1c2780999e2b9b22e3b3a540bed72a04cf803 has been merged

**Comment by tynes** [comment]
@boymanjor Has been rebased to remove intermediate commits
